### DISCUSSION TITLE
RHF is triggering 3 times as many rerenders on validation

### DISCUSSION THREAD

**Discussion by voinik** [react-hook-form/react-hook-form#11780] [Q&A]
### Version Number

7.51.1 and 7.51.2 (latest)

### Codesandbox/Expo snack

[This link](https://codesandbox.io/p/devbox/flw6cn?file=%2Fpackage.json%3A20%2C31&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cluqvxj3i00063b6fm4v7py4x%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cluqvxj3h00023b6fkfn3ueo6%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cluqvxj3h00043b6fuob4u434%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cluqvxj3h00053b6f9sjpakoo%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cluqvxj3h00023b6fkfn3ueo6%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluqvxj3h00013b6fdiunjsw6%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A20%252C%2522startColumn%2522%253A31%252C%2522endLineNumber%2522%253A20%252C%2522endColumn%2522%253A31%257D%255D%257D%252C%257B%2522id%2522%253A%2522cluqwu13i00023b6fcd1rdz4f%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A78%252C%2522startColumn%2522%253A20%252C%2522endLineNumber%2522%253A78%252C%2522endColumn%2522%253A20%257D%255D%252C%2522filepath%2522%253A%2522%252Ftailwind.config.ts%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522cluqvxj3h00023b6fkfn3ueo6%2522%252C%2522activeTabId%2522%253A%2522cluqvxj3h00013b6fdiunjsw6%2522%257D%252C%2522cluqvxj3h00053b6f9sjpakoo%2522%253A%257B%2522id%2522%253A%2522cluqvxj3h00053b6f9sjpakoo%2522%252C%2522activeTabId%2522%253A%2522cluqx7azi00yp3b6fzb4vlems%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522Development%2522%252C%2522port%2522%253A5173%252C%2522id%2522%253A%2522cluqx7azi00yp3b6fzb4vlems%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522cluqvxj3h00043b6fuob4u434%2522%253A%257B%2522id%2522%253A%2522cluqvxj3h00043b6fuob4u434%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522CSB_RUN_OUTSIDE_CONTAINER%253D1%2520devcontainer%2520templates%2520apply%2520--template-id%2520%255C%2522ghcr.io%252Fdevcontainers%252Ftemplates%252Ftypescript-node%255C%2522%2520--template-args%2520%27%257B%257D%27%2520--features%2520%27%255B%255D%27%2522%252C%2522id%2522%253A%2522cluqx55jr00mo3b6f6cellepp%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522id%2522%253A%2522cluqxxqsx01fo3b6feztwd3em%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cluqxxqwu000qdbhm1y2k4ic0%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Development%2522%252C%2522id%2522%253A%2522cluqy930w00kf3b6fhqh1rv9n%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cluqy930w00kf3b6fhqh1rv9n%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

### Steps to reproduce

First of all, thank you very much for this library. It has been my go-to for a long time now!

## Reproduction
1. Go to [this codesandbox devbox](https://codesandbox.io/p/devbox/flw6cn?file=%2Fpackage.json%3A20%2C31&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cluqvxj3i00063b6fm4v7py4x%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cluqvxj3h00023b6fkfn3ueo6%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cluqvxj3h00043b6fuob4u434%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cluqvxj3h00053b6f9sjpakoo%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cluqvxj3h00023b6fkfn3ueo6%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluqvxj3h00013b6fdiunjsw6%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A20%252C%2522startColumn%2522%253A31%252C%2522endLineNumber%2522%253A20%252C%2522endColumn%2522%253A31%257D%255D%257D%252C%257B%2522id%2522%253A%2522cluqwu13i00023b6fcd1rdz4f%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A78%252C%2522startColumn%2522%253A20%252C%2522endLineNumber%2522%253A78%252C%2522endColumn%2522%253A20%257D%255D%252C%2522filepath%2522%253A%2522%252Ftailwind.config.ts%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522cluqvxj3h00023b6fkfn3ueo6%2522%252C%2522activeTabId%2522%253A%2522cluqvxj3h00013b6fdiunjsw6%2522%257D%252C%2522cluqvxj3h00053b6f9sjpakoo%2522%253A%257B%2522id%2522%253A%2522cluqvxj3h00053b6f9sjpakoo%2522%252C%2522activeTabId%2522%253A%2522cluqx7azi00yp3b6fzb4vlems%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522Development%2522%252C%2522port%2522%253A5173%252C%2522id%2522%253A%2522cluqx7azi00yp3b6fzb4vlems%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522cluqvxj3h00043b6fuob4u434%2522%253A%257B%2522id%2522%253A%2522cluqvxj3h00043b6fuob4u434%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522CSB_RUN_OUTSIDE_CONTAINER%253D1%2520devcontainer%2520templates%2520apply%2520--template-id%2520%255C%2522ghcr.io%252Fdevcontainers%252Ftemplates%252Ftypescript-node%255C%2522%2520--template-args%2520%27%257B%257D%27%2520--features%2520%27%255B%255D%27%2522%252C%2522id%2522%253A%2522cluqx55jr00mo3b6f6cellepp%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522id%2522%253A%2522cluqxxqsx01fo3b6feztwd3em%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cluqxxqwu000qdbhm1y2k4ic0%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Development%2522%252C%2522id%2522%253A%2522cluqy930w00kf3b6fhqh1rv9n%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cluqy930w00kf3b6fhqh1rv9n%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
2. Open the console in the virtual browser on the right (click on "Preview console" in the top right, and then on the "Console" tab)
3. Clear the console for clarity
4. Click on the "Submit" button of the form in the browser
5. Notice the number of entries in the Console went from 0 (because you cleared it) to 6

If I downgrade the react-hook-form version from 7.51.1 to 7.51.0, then you will only see 2 entries in the Console. This means there is some kind of regression. I spent quite a bit digging but I'm not sure what's causing it yet. I noticed the release notes of 7.51.1 mentioned validatingFields. That seems to be related.

I commented out some other components that I was testing with, but they don't seem to be affected (there seems to be no difference between the old and newer versions of react-hook-form, with relation to rerenders).

Also, if you fill in the complete form correctly, and then submit, you will see 4 entries **before** the `onSubmit` `console.log(data)` call, and 4 entries **after** it, whereas with 7.51.0 you'll **only** see the `onSubmit` `console.log(data)`, and then 4 entries **after** it.

## Repo for testing
If you want to quickly play around with this, here's a repo I set up with create-next-app: https://github.com/voinik/rhf-rerender-issue

### Expected behaviour

I expect no extra rerenders compared to 7.51.0.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
thanks for the detailed issue report, i don't think we will try to eliminate the re-render, rather reduce the amount. i haven't looked close enough, but some minor extra re-render shouldn't be the root cause of your app perf issue.

**Comment by voinik** [comment]
No problem!
I agree that eliminating the rerenders isn't feasible. Reducing them to what they were in 7.51.0 was the goal of this report.

And indeed, it's definitely not a bottleneck for performance. I just thought you'd like to know as performance and minimizing rerenders are some of the unique selling points of this library. I'll of course leave it to you if it's worth fixing :)

**Comment by bluebill1049** [comment]
> And indeed, it's definitely not a bottleneck for performance. I just thought you'd like to know as performance and minimizing rerenders are some of the unique selling points of this library. I'll of course leave it to you if it's worth fixing :)

thank you for your understanding, I will take a look at this when i get my hands on it.

**Comment by vgoreev** [comment]
For some reason, accessing a Proxy object / submitting a form causing each input inside FormProvider to rerender exactly 3 times. Not accessing formState: { errors } on a Form level reduces form element rerenders but I'm still having field rerendering issue. Downgrading doesn't help as well

**Comment by bluebill1049** [comment]
could you double check if you are not subscribed to `validateFields`,  we have regression test for render count which is not affected: https://github.com/react-hook-form/react-hook-form/blob/master/cypress/e2e/basic.cy.ts#L100

File: https://github.com/react-hook-form/react-hook-form/blob/master/cypress/e2e/basic.cy.ts#L100

**Reply by voinik** [reply]
I think you mean `validatingFields`? But no, it's not being used in my example repo 😕

**Reply by voinik** [reply]
I'm going to dig a bit, but shadcn is using this in all form field components:
```tsx
  const { getFieldState, formState } = useFormContext()
```

**Comment by voinik** [comment]
Dug around a bit and I'm closer. It's happening in `src/logic/createFormControl.ts`, specifically here:
```ts
    const _executeSchema = async (name) => {
        _updateIsValidating(name, true); // <-- HERE
        const result = await _options.resolver(_formValues, _options.context, getResolverOptions(name || _names.mount, _fields, _options.criteriaMode, _options.shouldUseNativeValidation));
        _updateIsValidating(name); // <-- AND HERE
        return result;
    };
```

If I remove the lines I commented (they didn't exist before the PR I mention below), the additional rerenders no longer happen.

However, in the PR that introduced this change (#11624), you went from handling schemas in a synchronous way to an asynchronous way. I assume you added the `_updateIsValidating(..)` to keep track of the state of the validation?

Perhaps we should just take the rerender hit, and leave it as is? Or is there another way to deal with this?
